Implement post editing / edit history

zeitschlag opened the pull request:
# Rationale

| Demo | Description |
|---|---|
| ![RocketSim_Recording_iPhone_14_Pro_2023-02-23_09 52 04](https://user-images.githubusercontent.com/126418/220861284-c1d155a1-c35b-47c2-bd77-3b99aa2293af.gif) | Implement post-editing feature and edit history. Demo shows editing the status text and removing an image. |

zeitschlag committed:
Refactor status-metric to show entries below another (IOS-2)

DK1DA commented on Mastodon.xcworkspace/xcshareddata/IDETemplateMacros.plist:
```diff
@@ -0,0 +1,3 @@
+{
+	FILEHEADER = " Copyright © ___YEAR___ Mastodon gGmbH. All rights reserved.";
```
❤️  so much 👍

zeitschlag replied:
@kimar We could cherrypick it to develop? So others (and you) can profit from it, too?

zeitschlag committed:
Add some styling to rows (IOS-2)

zeitschlag committed:
Add (missing) relationship between status and application (IOS-2)

zeitschlag committed:
Show applicatoin-name (IOS-2)

zeitschlag committed:
Sprinkle in some localization (IOS-2)

zeitschlag committed:
Fix wrong type

zeitschlag committed:
Add new model version (IOS-2)

includes first draft of StatusEdit, I need to doublecheck the relationships

zeitschlag committed:
Download StatusEdits (IOS-2)

j-f1 commented on MastodonSDK/Sources/MastodonUI/View/Content/StatusMetricRowView.swift:
```diff
@@ -0,0 +1,73 @@
+import UIKit
+import MastodonAsset
+
+public final class StatusMetricRowView: UIButton {
```
If possible it would be nice to have this as a separate table view cell so that it can take advantage of system behaviors like rendering the selection highlight.

zeitschlag replied:
Noted, thanks for your feedback. This is a work in progress and this was the faster way to achieve the desired result. I'll have a look into it.

zeitschlag committed:
Show edit-history (IOS-2)

zeitschlag committed:
Add base for StatusEdit-persistence (IOS-2)

I hate Core Data from the bottom of my heart 🥲

zeitschlag committed:
Load and show edit history (IOS-2)

zeitschlag committed:
Show edit-history in status view (IOS-2)

zeitschlag committed:
Kill separator (IOS-2)

zeitschlag committed:
Format the date (IOS-2)

zeitschlag committed:
Show last edit-date (IOS-2)

zeitschlag committed:
Set colors and fonts (IOS-2)

zeitschlag committed:
Decrease font-size (IOS-2)

zeitschlag committed:
Mark original post (IOS-2)

zeitschlag committed:
Add emojis (IOS-2)

zeitschlag committed:
Fix typo (IOS-2)

zeitschlag committed:
Remove all status-edits for a status when updating (IOS-2)

Consider this a first draft :see_no_evil:

zeitschlag committed:
Add edit-entry to status-menu (IOS-2)

zeitschlag committed:
Put Delete into its own submenu (IOS-2)

zeitschlag committed:
Add edit-icon for status-view (IOS-2)

zeitschlag committed:
Add and use editedAt-date to status (IOS-2)

Why didn't I see this before?! I guess, I need some glasses.

Oh, and of course we use this new property. If there's no editedAt-date, then there's no button.

zeitschlag committed:
Fix language-selection-crash (IOS-2)

zeitschlag committed:
Show different buttons when editing a status (IOS-2)

zeitschlag committed:
Inject status into viewmodels for editing (IOS-2)

zeitschlag committed:
[WIP] Send status-edits (IOS-2)

It doesn't work yet, server responds with 500.

zeitschlag committed:
Download statusSource (IOS-2)

zeitschlag committed:
Send JSON instead of multipart-form-stuff (IOS-2)

I just edited a toot successfully for the first time.

zeitschlag committed:
Download status after editing it (IOS-2)

zeitschlag committed:
Add some localization (IOS-2)

zeitschlag committed:
Add some documentation (IOS-2)

kimar committed:
Don't show translate activity indicator when trying to edit a post (IOS-2)

kimar committed:
Reload Home Timeline after editing post (IOS-2)

kimar committed:
Implement status reload after editing in thread (IOS-2)

kimar committed:
Implement I18n for new thread view items (IOS-2)

kimar committed:
Implement "Edit X" text for Status (IOS-2)

kimar committed:
Fix edit post functionality on iPad (IOS-2)

kimar committed:
Fix CW discarded/ignored when editing (IOS-2)

kimar committed:
Implement editing Poll (IOS-2)

kimar committed:
Safeguard poll without status (IOS-2)

kimar committed:
Move CoreData Model changes to Model v9 (IOS-2)

DK1DA commented on MastodonSDK/Sources/CoreDataStack/CoreData.xcdatamodeld/CoreData 9.xcdatamodel/contents:
```diff
@@ -0,0 +1,291 @@
+        <attribute name="votersCount" optional="YES" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="votesCount" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <relationship name="options" toMany="YES" deletionRule="Cascade" destinationEntity="PollOption" inverseName="poll" inverseEntity="PollOption"/>
+        <relationship name="status" optional="YES" maxCount="1" deletionRule="Nullify" destinationEntity="Status" inverseName="poll" inverseEntity="Status"/>
```
`status` on `Poll` is now optional as otherwise there are issues when editing a status and deleting a post.

j-f1 commented:
Since presumably only you and a couple of people ever installed an app using CoreData 8, would it be possible to make the changes to CoreData 8 directly instead of making a new version? (I think it’s also fine to change the optional status of a field without making a new version)

kimar committed:
Lock visibility on edit post (IOS-2)

kimar committed:
Implement editable attachments (IOS-2)

kimar committed:
Fix StatusEdit can't be decoded because Poll model won't match API (IOS-2)

kimar committed:
Fix edit history not showing poll (IOS-2)

kimar committed:
Fix media not show in edit history (IOS-2)

kimar committed:
Improve layout for StatusView in StatusEditHistoryTableViewCell (IOS-2)

kimar committed:
Don't show favorites/reblog button in thread if no favs/reblogs exist (IOS-2)

kimar committed:
Fetch edit history after publishing an edit (IOS-2)

kimar committed:
Improve StatusEditHistoryTableViewCell and fix iPad margin issues (IOS-2)

kimar committed:
Fix missing constraints edit history layout issues (IOS-2)

zeitschlag committed:
Add fix for #940 (IOS-2)

kimar committed:
Fix MetaTextViewRepresentable height is not correct when editing posts (IOS-2)

kimar commented:
@j-f1 I was also considering this, I'll make the changes t model v8 directly then.

kimar committed:
Revert "Move CoreData Model changes to Model v9 (IOS-2)"

This reverts commit 753e2ca4ed9b50ea3c9897cc02b5d7effe399da9.

kimar committed:
Revert spaces (IOS-2)

j-f1 commented on MastodonSDK/Sources/CoreDataStack/Entity/Mastodon/Poll.swift:
```diff
@@ -31,14 +31,14 @@ public final class Poll: NSManagedObject {
     // sourcery: autoGenerateProperty
     @NSManaged public private(set) var createdAt: Date
-    // sourcery: autoUpdatableObject, autoGenerateProperty
+    // sourcery: autoUpdatableObject, autoGenerateProperty  
```
```suggestion
    // sourcery: autoUpdatableObject, autoGenerateProperty
```

zeitschlag replied:
It's just the whitespace, right?

j-f1 replied:
Yep.

j-f1 reviewed (approved):
This looks awesome! I opened https://github.com/mastodon/mastodon-ios/pull/951 with a few improvements; namely, attempting to emulate the styling of table view cells for `StatusMetricRowView`, improving display at large text sizes, and improving VoiceOver accessibility. It’s not currently possible to access the edit history via VoiceOver (underlying issue for that is IOS-70 which I will try to handle today or tomorrow).

zeitschlag commented on Mastodon/Scene/Compose/ComposeViewController.swift:
```diff
@@ -141,6 +189,14 @@ extension ComposeViewController {
         present(alertController, animated: true, completion: nil)
     }
 
+    private var rightBarButtonItemForCurrentContext: UIBarButtonItem {
+        switch viewModel.composeContext {
+        case .composeStatus:
+            return publishBarButtonItem
+        case .editStatus:
+            return saveBarButtonItem
+        }
+    }
```
That's neat, I like this!

zeitschlag commented on Mastodon/Scene/HomeTimeline/HomeTimelineViewController.swift:
```diff
@@ -200,6 +200,12 @@ extension HomeTimelineViewController {
             .store(in: &disposeBag)
         
+        context.publisherService.statusPublishResult.sink { result in
+            if case let Result.success(res) = result, case StatusPublishResult.edit = res {
```
I've never seen a nested `if case let, case`, but it totally makes sense 👍

j-f1 replied:
I think this could also be 

```suggestion
            if case .success(.edit) = result {
```

(but the 2 should be equivalent)

zeitschlag commented on Mastodon/Scene/Thread/Edit History/StatusEditHistoryTableViewCell.swift:
```diff
@@ -0,0 +1,134 @@
+    }
+}
+
+class StatusHistoryView: UIView {
```
More of a suggestion: We could refactor this into its own file. Feel free, not to :)

zeitschlag commented on Mastodon/Scene/Thread/Edit History/StatusEditHistoryTableViewCell.swift:
```diff
@@ -0,0 +1,134 @@
+class StatusHistoryView: UIView {
+    let statusView = StatusView()
+    
+    private var statusViewLeadingConstraint: NSLayoutConstraint!
```
The reason for force unwrapping (`NSLayoutConstraint!`) is that you can only set the constraints after adding `statusView` as a subview in `init`, right? What's your opinion on `NSLayoutConstraint?` instead?

DK1DA replied:
Exactly, but we can easily remove the implicit unwrap, I've done this now 👍

zeitschlag committed:
Add some localization (IOS-2)

zeitschlag commented on Mastodon/Scene/Thread/Edit History/StatusEditHistoryViewController.swift:
```diff
@@ -0,0 +1,80 @@
+            let dateText: String
+
+            if statusEdit == viewModel.edits.last {
+                dateText = L10n.Common.Controls.Status.EditHistory.originalPost(self.dateFormatter.string(from: statusEdit.createdAt))
```
I added some L10n here.

zeitschlag commented on Mastodon/Scene/Thread/Edit History/StatusEditHistoryViewController.swift:
```diff
@@ -0,0 +1,80 @@
+        view.backgroundColor = ThemeService.shared.currentTheme.value.secondarySystemBackgroundColor
+        setupConstraints()
+
+        title = L10n.Common.Controls.Status.EditHistory.title
```
and here, too.

DK1DA replied:
Thx!

zeitschlag commented on MastodonSDK/Sources/MastodonCore/Extension/CoreDataStack/StatusEdit+Property.swift:
```diff
@@ -0,0 +1,16 @@
+            content: entity.content,
+            sensitive: entity.sensitive,
+            spoilerText: entity.spoilerText,
+            emojis: entity.mastodonEmojis)
```
Emojis are there, but I can't explain, why `poll` and `mediaAttachements` are missing here, at least according to [StatusEdit](https://docs.joinmastodon.org/entities/StatusEdit/) they should be there? Am I missing something?

DK1DA replied:
Hm yeah, weird 😕  I just added `mediaAttachments` and will add poll now...

zeitschlag commented on MastodonSDK/Sources/MastodonCore/Persistence/Persistence+StatusEdit.swift:
```diff
@@ -0,0 +1,58 @@
+        in managedObjectContext: NSManagedObjectContext,
+        forStatusID statusID: Status.ID
+    ) -> [StatusEdit] {
+        //        let request = StatusEditHistoryEntry.fetchRequest()
+        //        let statusEdit = try? managedObjectContext.fetch(request).first as? StatusEditHistoryEntry
+        //        return statusEdit
```
This doesn't look quite right, yet 🤨

zeitschlag committed:
Fix typo (IOS-2)

zeitschlag committed:
Remove comments (IOS-2)

zeitschlag committed:
Localize menu-entry for post-edit (IOS-2)

zeitschlag reviewed (commented):
Most of the stuff looks good to me, although I'm a bit uncertain about `Persistence+StatusEdit.swift`.

j-f1 committed:
Improvements to StatusMetricRowView / edit history (#951)

zeitschlag commented on MastodonSDK/Sources/MastodonUI/View/Content/StatusMetricRowView.swift:
```diff
@@ -0,0 +1,119 @@
+            detailLabel.textAlignment = .natural
+        } else {
+            contentStack.axis = .horizontal
+            contentStack.alignment = .leading
```
```suggestion
            contentStack.alignment = .center
```

![stackview-center](https://user-images.githubusercontent.com/2580019/221036642-86dcb206-def8-4c47-b1b3-809a4d957250.jpeg)

j-f1 replied:
Will add that, working on a couple of fixes here too! (the detail should be on  line)

zeitschlag replied:
Don't worry, it's basically already taken care of with this nice suggestion-feature :)

j-f1 replied:
Done in #953, also fixed the issue where the text would break onto 2 lines even if there was enough space for it to be on 1

DK1DA replied:
Thank you @j-f1 ! 👏

kimar committed:
Fix status history is not reflecting media changes (IOS-2)

kimar committed:
Update MastodonSDK/Sources/CoreDataStack/Entity/Mastodon/Poll.swift

Co-authored-by: Jed Fox <git@jedfox.com>

kimar committed:
Update Mastodon/Scene/HomeTimeline/HomeTimelineViewController.swift

Co-authored-by: Jed Fox <git@jedfox.com>

kimar committed:
Move StatusHistoryView into it's own file (IOS-2)

kimar committed:
Don't implicitly unwrap constraint (IOS-2)

kimar committed:
Remove unused code (IOS-2)

j-f1 committed:
IOS-70: Improve StatusMetricView Accessibility (#953)

kimar committed:
Add poll history model to coredata (IOS-2)

kimar committed:
Implement Poll history (IOS-2)

kimar committed:
Merge branch 'develop' into 2-edit-post

zeitschlag commented on Mastodon/Scene/Share/View/Content/PollOptionView+Configuration.swift:
```diff
@@ -109,3 +114,30 @@ extension PollOptionView {
+            .sink { [weak self] theme in
+                guard let self = self else { return }
+                self.checkmarkBackgroundView.backgroundColor = UIColor(dynamicProvider: { trailtCollection in
+                    return trailtCollection.userInterfaceStyle == .light ? .white : theme.tableViewCellSelectionBackgroundColor
```
Is there a reason why we use `.white` for `.light` instead of relying on the asset-catalog-color?

DK1DA replied:
Good question this code has been in place for a while and I'm not able to reason about it, would need to further investigate this.

kimar committed:
Implement poll not reflecting edits in existing status (IOS-2)

kimar committed:
Fix poll order while / after editing (IOS-2)

kimar committed:
Fix poll options are duplicated and not properly updated (IOS-2)

kimar committed:
Improve content warning display for edit history (IOS-2)

kimar committed:
Fix CW toggle visible in some history entries (IOS-2)

kimar committed:
Merge branch 'develop' into 2-edit-post

zeitschlag committed:
Merge branch 'develop' into 2-edit-post
